### PR TITLE
MULE-11440: Provide a way to set the SecurityContext in SDK

### PR DESCRIPTION
**Pull request by alegmarra**

### CODE REVIEW

File: src/main/java/org/mule/runtime/api/security/Authentication.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * <code>Authentication</code> represents an authentication request and contains authentication information
+ * if the request was successful
+ *
+ * @since 1.0
+ */
+public interface Authentication extends Serializable {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/Authentication.java]: jdoc

File: src/main/java/org/mule/runtime/api/security/Authentication.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * <code>Authentication</code> represents an authentication request and contains authentication information
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/Authentication.java]: `<code>` -> `{@code}`

File: src/main/java/org/mule/runtime/api/exception/ErrorMessageAwareException.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.exception;
+
+import org.mule.runtime.api.message.Message;
+
+/**
+ * Exception that holds a {@link Message} regarding the error that occurred.
+ * When this exception is analysed to create an {@link org.mule.runtime.api.message.Error}, it will feature the message as it's
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/exception/ErrorMessageAwareException.java]: import

File: src/main/java/org/mule/runtime/api/security/SecurityContext.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import java.io.Serializable;
+
+/**
+ * <code>SecurityContext</code> holds security information and is associated with the MuleSession.
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/SecurityContext.java]: code

File: src/main/java/org/mule/runtime/api/security/SecurityException.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.i18n.I18nMessage;
+
+/**
+ * <code>SecurityException</code> is a generic security exception
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/SecurityException.java]: code

File: src/main/java/org/mule/runtime/api/security/UnauthorisedException.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import org.mule.runtime.api.i18n.I18nMessage;
+
+/**
+ * <code>UnauthorisedException</code> is thrown if authentication fails.
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/UnauthorisedException.java]: code

File: src/main/java/org/mule/runtime/api/security/UnknownAuthenticationTypeException.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import org.mule.runtime.api.exception.MuleException;
+
+/**
+ * <code>UnknownAuthenticationTypeException</code> is thrown if a security context request is make with an unrecognised
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/UnknownAuthenticationTypeException.java]: code

File: src/main/java/org/mule/runtime/api/security/UnsupportedAuthenticationSchemeException.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.security;
+
+import org.mule.runtime.api.exception.ErrorMessageAwareException;
+import org.mule.runtime.api.i18n.I18nMessage;
+import org.mule.runtime.api.message.Message;
+
+/**
+ * <code>UnsupportedAuthenticationSchemeException</code> is thrown when a authentication scheme is being used on the message that
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/UnsupportedAuthenticationSchemeException.java]: code

File: src/main/java/org/mule/runtime/api/security/SecurityContext.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ * 
+ * @since 1.0
+ */
+public interface SecurityContext extends Serializable {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/SecurityContext.java]: I don't really see the need for this class to be API. All we need is to be able to get and set the Authentication object and providers. I don't see what an extensions user would do with this

File: src/main/java/org/mule/runtime/api/security/Authentication.java (hunk 1 above)
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/security/Authentication.java]: how does the extension developer create an instance of this class? Implementations are all in non-api core

File: src/main/java/org/mule/runtime/api/exception/ErrorMessageAwareException.java (hunk 3 above)
Comment (alegmarra) [inline src/main/java/org/mule/runtime/api/exception/ErrorMessageAwareException.java]: can't

**Pull request merged**